nix/postgres/setup-script.nix

```diff
@@ -101,7 +100,7 @@ in
       echo
       echo "PostgreSQL is setting up the initial database."
       echo
-      PGHOST=$(mktemp -d "$(readlink -f ${config.dataDir})/pg-init-XXXXXX")
+      PGHOST=$(mkdir -p "${config.socketDir}" && mktemp -d "$(readlink -f "${config.socketDir}")/pg-init-XXXXXX")
```

Comment:
Why do we have to `mkdir` here? wouldn’t just `mktemp` suffice? because we are using this only to temporarily start the server during initialisation and we delete the socket after init.

Reply:
You are correct; this was left over from a previous command modification and is no longer needed.
Now to try and figure out why the CI test seems to hang and fix that as well.

Reply:
The CI test that failed earlier fails on what looks like a race condition for the postgres test when both `./data/pg1` and `./data/pg2` are created at the same time, and one of the two fails to start because of a "File exists" error.  Ie:
```
default-test> [pg2-init ] initdb: error: could not create directory "./data": File exists
default-test> [pg1-init ] creating directory ./data/pg1 ... ok
default-test> [pg2-init ] creating directory ./data/pg2 ...
...
<pg2 then fails to start in this example>
```

I can reproduce the failure locally with nixci, but I cannot reproduce it by executing the commands manually outside of the nixci test script, or by loading the same test setup into a process-compose yaml and running it.

We could change the postgresql_test.nix file to use two separate parent dataDirs which resolves this race condition test failure, like:
```
  services.postgres."pg1" = {
    enable = true;
+   dataDir = "./data1/pg1";
    ...
  };
  services.postgres."pg2" = {
    enable = true;
+   dataDir = "./data2/pg2";
    ...
  };
```

When I test with nixci locally using the main services-flake branch, it also fails this way.  Strangely, the CI history doesn't seem to show it failing like this in the past.  Have you encountered this before?

Reply:
Actually, yes. This is the reason for https://github.com/juspay/services-flake/issues/74


I have also raised a fix for this: https://github.com/juspay/services-flake/pull/78

Now the CI is failing due to an unrelated error: “Error: devour-flake produced an outpath (/nix/store/1s02vc584sy7nzhx81j93mk1mrwnql9g-devour-output) with no outputs”

Reply:
And this seems to be happening consistently across all the CI runs. I can’t reproduce it locally either.

Reply:
> Now the CI is failing due to an unrelated error: “Error: devour-flake produced an outpath (/nix/store/1s02vc584sy7nzhx81j93mk1mrwnql9g-devour-output) with no outputs”

Looks like Nix 2.19 had a regression. See https://github.com/srid/devour-flake/issues/22

Is there an official bug report? Or was there a backwards incompatible change?

Reply:
Should be fixed (in our repo) after https://github.com/juspay/services-flake/pull/79

Reply:
@shivaraj-bh, apologies for revisiting this, but back to the original question of why do we have mkdir here:
```
-     PGHOST=$(mktemp -d "$(readlink -f ${config.dataDir})/pg-init-XXXXXX")
+     PGHOST=$(mkdir -p "${config.socketDir}" && mktemp -d "$(readlink -f "${config.socketDir}")/pg-init-XXXXXX")
```

When configDir and socketDir are the same (the default), the `initDb -D ${config.dataDir}` earlier in the postgres setup-script does the equivalent of a `mkdir -p` for config.dataDir which is why mktemp in that case doesn't require it.

It's also not a problem when configDir and socketDir are different and the full path for the socketDir already exists, no mkdir -p is required.

In the remaining case where configDir and socketDir are different and socketDir path doesn't exist yet or only exists partially, mktemp only tries to create the leaf node directory, but not any other part of the parental dir path and so it fails.
```
❯ mktemp -d 1/2/3/pg-init-XXXXXX
mktemp: failed to create directory via template ‘1/2/3/pg-init-XXXXXX’: No such file or directory

❯ mkdir -p 1/2/3 && mktemp -d 1/2/3/pg-init-XXXXXX
1/2/3/pg-init-Vwl7c7
```

When I commented yesterday it wasn't needed anymore, I was no longer testing this 3rd case and didn't realize it was still an issue, but I ran into it again today.  I'll PR a line line update shortly.

Reply:
Makes sense!